core/cache.hpp

```diff
@@ -0,0 +1,149 @@
+// Copyright 2016 Husky Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <list>
+#include <unordered_map>
+#include <utility>
+
+#include "boost/optional.hpp"
+
+namespace husky {
+
+template <typename KeyT, typename ValT>
+class ICache {
+   public:
+    explicit ICache(size_t size) { size_ = size; }
+
+    size_t get_max_size() { return size_; }
+
+   protected:
+    size_t size_;
+};
+
+template <typename KeyT, typename ValT>
+class LRUCache : public ICache<KeyT, ValT> {
+   public:
+    using ListIterator = typename std::list<std::pair<KeyT, ValT>>::iterator;
+
+    explicit LRUCache(size_t size) : ICache<KeyT, ValT>(size) {}
+
+    ~LRUCache() = default;
+
+    template <typename KeyU, typename ValU>
```

Comment:
Since if the function wants to deal with two situations where the argument is a lvalue or a rvalue, then it may be done by providing two overloading functions with different parameters(one const &, and one &&), or be done by providing a universal preference parameter which deals with these two situations at the same time. If we do it in the first way, then we can add virtual functions 'put' 'del' etc. in the base class to enforce the interfaces. If we do it in the second way, then we cannot override declare function templates as virtual in the base class, therefore being not able to enforce the interfaces. I may change to the first way